spec/controllers/protected_resources_controller_spec.rb

```diff
@@ -174,4 +174,106 @@ def doorkeeper_unauthorized_render_options(error: nil)
       end
     end
   end
+
+  context 'when custom forbidden render options are configured' do
+    before do
+      expect(Doorkeeper::AccessToken).to receive(:by_token).with(token_string).and_return(token)
+      expect(token).to receive(:acceptable?).with([:write]).and_return(false)
+    end
+
+    after do
+      module ControllerActions
+        def doorkeeper_forbidden_render_options(*)
+        end
+      end
+    end
+
+    controller do
+      before_filter -> { doorkeeper_authorize! :write }
+
+      include ControllerActions
+    end
+
+    let(:token) do
+      double(Doorkeeper::AccessToken,
+             accessible?: true, scopes: ['public'], revoked?: false, expired?: false)
+    end
+    let(:token_string) { '1A2DUWE' }
+
+    context 'with a JSON custom render' do
+      before do
+        module ControllerActions
+          def doorkeeper_forbidden_render_options(*)
+            { json: { error_message: 'Forbidden' } }
+          end
+        end
+      end
+
+      it 'renders a custom JSON response' do
+        forbidden_render_test
+        expect(response.content_type).to eq('application/json')
+        expect(response.status).to eq 403
+        parsed_body = JSON.parse(response.body)
+        expect(parsed_body).not_to be_nil
+        expect(parsed_body['error_message']).to match('Forbidden')
+      end
+    end
+
+    context 'with a status and JSON custom render' do
+      before do
+        module ControllerActions
+          def doorkeeper_forbidden_render_options(*)
+            { json: { error_message: 'Not Found' },
+              not_found_for_forbidden: true }
+          end
+        end
+      end
+
+      it 'renders a custom JSON response' do
+        forbidden_render_test
+        expect(response.content_type).to eq('application/json')
+        expect(response.status).to eq 404
+        parsed_body = JSON.parse(response.body)
+        expect(parsed_body).not_to be_nil
+        expect(parsed_body['error_message']).to match('Not Found')
+      end
+    end
+
+    context 'with a text custom render' do
+      before do
+        module ControllerActions
+          def doorkeeper_forbidden_render_options(*)
+            { text: 'Forbidden' }
+          end
+        end
+      end
+
+      it 'renders a custom status code and text response' do
+        forbidden_render_test
+        expect(response.status).to eq 403
+        expect(response.body).to eq('Forbidden')
+      end
+    end
+
+    context 'with a status and text custom render' do
+      before do
+        module ControllerActions
+          def doorkeeper_forbidden_render_options(*)
+            { not_found_for_forbidden: true, text: 'Not Found' }
+          end
+        end
+      end
+
+      it 'renders a custom status code and text response' do
+        forbidden_render_test
```

Comment:
I wouldn't extract a `forbidden_render_test` method. It's body is key to what's happening in the test, it reads better if we inline it so we don't need to jump back and forth to grasp the whole story.